File: backend/app/blocks.py
from . import modules

SIZE_COMPACT = "compact"
SIZE_NORMAL = "normal"
SIZES = (SIZE_COMPACT, SIZE_NORMAL)

AREA_OVERVIEW = "overview"
AREA_TIMETABLE = "timetable"
AREA_ABSENCE = "absence"
AREA_POST = "post"
AREA_MESSENGER = "messenger"
AREA_CONFERENCES = "conferences"
AREA_MODULES = {
    AREA_TIMETABLE: (modules.TIMETABLE,),
    AREA_ABSENCE: (modules.ABSENCES,),
    AREA_POST: (modules.LETTERS, modules.PINBOARD),
    AREA_MESSENGER: (modules.MESSENGER,),
    AREA_CONFERENCES: (modules.CONFERENCES,),
}
DEFAULT_NAVIGATION = (AREA_TIMETABLE, AREA_ABSENCE, AREA_POST, AREA_MESSENGER, AREA_CONFERENCES)

BLOCKS = (
    {"key": "today", "module": modules.TIMETABLE, "area": AREA_TIMETABLE, "compact": 3, "normal": 10, "size": SIZE_NORMAL},
    {"key": "next_lesson", "module": modules.TIMETABLE, "area": AREA_TIMETABLE, "compact": 1, "normal": 1, "size": SIZE_COMPACT},
    {"key": "week", "module": modules.TIMETABLE, "area": AREA_TIMETABLE, "compact": 1, "normal": 1, "size": SIZE_NORMAL},
    {"key": "letters", "module": modules.LETTERS, "area": AREA_POST, "compact": 3, "normal": 5, "size": SIZE_NORMAL},
    {"key": "noticeboard", "module": modules.PINBOARD, "area": AREA_POST, "compact": 3, "normal": 5, "size": SIZE_COMPACT},
    {"key": "absences", "module": modules.ABSENCES, "area": AREA_ABSENCE, "compact": 2, "normal": 5, "size": SIZE_NORMAL},
    {"key": "conferences", "module": modules.CONFERENCES, "area": AREA_CONFERENCES, "compact": 1, "normal": 3, "size": SIZE_NORMAL},
    {"key": "holidays", "module": modules.TIMETABLE, "area": AREA_TIMETABLE, "compact": 1, "normal": 3, "size": SIZE_COMPACT},
    {"key": "changes", "module": modules.TIMETABLE, "area": AREA_TIMETABLE, "compact": 3, "normal": 6, "size": SIZE_COMPACT},
    {"key": "chat", "module": modules.MESSENGER, "area": AREA_MESSENGER, "compact": 3, "normal": 5, "size": SIZE_COMPACT},
)
BLOCK_KEYS = tuple(block["key"] for block in BLOCKS)
BLOCK_BY_KEY = {block["key"]: block for block in BLOCKS}

DEFAULT_OVERVIEW_KEYS = ("today", "letters", "noticeboard", "conferences", "changes", "chat")


def module_of(key):
    return BLOCK_BY_KEY[key]["module"]


def default_overview_blocks():
    return [{"key": block["key"], "size": block["size"]} for block in BLOCKS if block["key"] in DEFAULT_OVERVIEW_KEYS]


def default_navigation():
    return list(DEFAULT_NAVIGATION)


def normalize_modules_disabled(raw):
    listed = raw if isinstance(raw, list) else []
    kept = []
    for entry in listed:
        name = str(entry) if isinstance(entry, str) else ""
        if name in modules.MODULES and name not in kept:
            kept.append(name)
    return kept


def normalize_overview_blocks(raw, disabled=()):
    if not isinstance(raw, list):
        return default_overview_blocks()
    off = set(disabled)
    kept = []
    seen = set()
    for entry in raw:
        key = entry.get("key") if isinstance(entry, dict) else entry
        if not isinstance(key, str) or key not in BLOCK_BY_KEY or key in seen:
            continue
        if BLOCK_BY_KEY[key]["module"] in off:
            continue
        size = entry.get("size") if isinstance(entry, dict) else None
        seen.add(key)
        kept.append({"key": key, "size": size if size in SIZES else BLOCK_BY_KEY[key]["size"]})
    return kept


def normalize_navigation(raw):
    listed = raw if isinstance(raw, list) else []
    kept = []
    for entry in listed:
        area = str(entry) if isinstance(entry, str) else ""
        if area in AREA_MODULES and area not in kept:
            kept.append(area)
    for area in DEFAULT_NAVIGATION:
        if area not in kept:
            kept.append(area)
    return kept


def area_enabled(area, available, disabled):
    return any(available.get(name, True) and name not in disabled for name in AREA_MODULES.get(area, ()))

File: frontend/blocks.js
(() => {
  const SIZE_COMPACT = "compact";
  const SIZE_NORMAL = "normal";
  const SIZES = [SIZE_COMPACT, SIZE_NORMAL];
  const AREA_OVERVIEW = "overview";
  const AREA_MODULES = {
    timetable: ["timetable"],
    absence: ["absences"],
    post: ["letters", "pinboard"],
    messenger: ["messenger"],
    conferences: ["conferences"],
  };
  const DEFAULT_NAVIGATION = ["timetable", "absence", "post", "messenger", "conferences"];
  const BAR_LIMIT = 5;
  const BLOCK_SEARCH_FROM = 20;
  const MORE_KEY = "more";

  const BLOCK_CATALOGUE = [
    { key: "today", module: "timetable", area: "timetable", compact: 3, normal: 10, size: "normal", surfaces: ["overview", "card"] },
    { key: "next_lesson", module: "timetable", area: "timetable", compact: 1, normal: 1, size: "compact", surfaces: ["overview", "card"] },
    { key: "week", module: "timetable", area: "timetable", compact: 1, normal: 1, size: "normal", surfaces: ["overview", "card"] },
    { key: "letters", module: "letters", area: "post", compact: 3, normal: 5, size: "normal", surfaces: ["overview", "card"] },
    { key: "noticeboard", module: "pinboard", area: "post", compact: 3, normal: 5, size: "compact", surfaces: ["overview", "card"] },
    { key: "absences", module: "absences", area: "absence", compact: 2, normal: 5, size: "normal", surfaces: ["overview", "card"] },
    { key: "conferences", module: "conferences", area: "conferences", compact: 1, normal: 3, size: "normal", surfaces: ["overview", "card"] },
    { key: "holidays", module: "timetable", area: "timetable", compact: 1, normal: 3, size: "compact", surfaces: ["overview", "card"] },
    { key: "changes", module: "timetable", area: "timetable", compact: 3, normal: 6, size: "compact", surfaces: ["overview", "card"] },
    { key: "chat", module: "messenger", area: "messenger", compact: 3, normal: 5, size: "compact", surfaces: ["overview"] },
  ];
  const BLOCK_BY_KEY = new Map(BLOCK_CATALOGUE.map((block) => [block.key, block]));
  const DEFAULT_OVERVIEW_KEYS = ["today", "letters", "noticeboard", "conferences", "changes", "chat"];

  function blockOf(key) {
    return BLOCK_BY_KEY.get(key) || null;
  }

  function blockKeys() {
    return BLOCK_CATALOGUE.map((block) => block.key);
  }

  function sizeOf(block, size) {
    return SIZES.includes(size) ? size : block.size;
  }

  function limitOf(block, size) {
    return block[sizeOf(block, size)];
  }

  function hasContent(items) {
    return Array.isArray(items) && items.length > 0;
  }

  function shownItems(block, size, items) {
    const list = Array.isArray(items) ? items : [];
    const limit = limitOf(block, size);
    return { items: list.slice(0, limit), more: list.length > limit };
  }

  function offeredBlocks(moduleOn, surface) {
    return BLOCK_CATALOGUE.filter((block) => moduleOn(block.module) && (!surface || block.surfaces.includes(surface)));
  }

  function defaultOverviewBlocks() {
    return BLOCK_CATALOGUE.filter((block) => DEFAULT_OVERVIEW_KEYS.includes(block.key)).map((block) => ({ key: block.key, size: block.size }));
  }

  function normalizeOverviewBlocks(raw, offered) {
    const allowed = new Set((offered || BLOCK_CATALOGUE).map((block) => block.key));
    const source = Array.isArray(raw) ? raw : defaultOverviewBlocks();
    const kept = [];
    const seen = new Set();
    for (const entry of source) {
      const key = entry && typeof entry === "object" ? entry.key : entry;
      const block = typeof key === "string" ? blockOf(key) : null;
      if (!block || seen.has(block.key) || !allowed.has(block.key)) continue;
      seen.add(block.key);
      kept.push({ key: block.key, size: sizeOf(block, entry && typeof entry === "object" ? entry.size : null) });
    }
    return kept;
  }

  function hiddenBlocks(enabled, offered) {
    const on = new Set(enabled.map((entry) => entry.key));
    return (offered || BLOCK_CATALOGUE).filter((block) => !on.has(block.key));
  }

  function normalizeNavigation(raw) {
    const kept = [];
    for (const entry of Array.isArray(raw) ? raw : []) {
      if (typeof entry === "string" && AREA_MODULES[entry] && !kept.includes(entry)) kept.push(entry);
    }
    for (const area of DEFAULT_NAVIGATION) {
      if (!kept.includes(area)) kept.push(area);
    }
    return kept;
  }

  function moveKey(list, key, direction) {
    const index = list.indexOf(key);
    const target = index + direction;
    if (index < 0 || target < 0 || target >= list.length) return list.slice();
    const moved = list.slice();
    moved.splice(index, 1);
    moved.splice(target, 0, key);
    return moved;
  }

  function barLayout(areas, limit) {
    const cap = Math.max(2, Number(limit) || BAR_LIMIT);
    const entries = [AREA_OVERVIEW].concat(areas);
    if (entries.length <= cap) return { bar: entries, more: [] };
    return { bar: entries.slice(0, cap - 1).concat([MORE_KEY]), more: entries.slice(cap - 1) };
  }

  window.RanzenpostBlocks = {
    SIZE_COMPACT,
    SIZE_NORMAL,
    SIZES,
    AREA_OVERVIEW,
    AREA_MODULES,
    DEFAULT_NAVIGATION,
    BAR_LIMIT,
    BLOCK_SEARCH_FROM,
    MORE_KEY,
    BLOCK_CATALOGUE,
    DEFAULT_OVERVIEW_KEYS,
    blockOf,
    blockKeys,
    sizeOf,
    limitOf,
    hasContent,
    shownItems,
    offeredBlocks,
    defaultOverviewBlocks,
    normalizeOverviewBlocks,
    hiddenBlocks,
    normalizeNavigation,
    moveKey,
    barLayout,
  };
})();
